### PR TITLE
refactor: rewrite music module

### LINKED ISSUE #1284: request: make music's queue be cached in Redis
**Is your feature request related to a problem? Please describe.**

Whenever Skyra reboots, the queue is lost, and that causes some frustration in users.

**Describe the solution you'd like**

Move the cache to be backed by Redis, this way it's not lost. One way to do this is by porting code from [`lavaqueue`](https://github.com/lavalibs/lavaqueue) in Skyra, as the official library doesn't support [`lavacord`](https://www.npmjs.com/package/lavacord).

**Describe alternatives you've considered**

N/a.

**Additional context**

N/a.

### LINKED ISSUE #1336: rewrite: music module improvements
At the moment, the music module is kind of fragile, it barely holds any state, and sometimes, it results on heavy annoyances when the package misbehaves (e.g. sending "Playing" messages non-stop in some hard-to-reproduce glitches).

I would like to change the music library again, this one, one of our own, made by and for Skyra Project, with the best of both libraries we used:

## [`lavalink`](https://www.npmjs.com/package/lavalink)

- 💚 Easy [`lavaqueue`](https://www.npmjs.com/package/lavaqueue) support (made by same maintainer).
- 💚 Handlers are very robust.
- 💚 Audio playback doesn't stop when Skyra restarts.
- 💚 Doesn't hold a lot of state.
- 🤍 Support exists.
- ❌ Implicit connection.
- ❌ Crash loop when the server is offline (restarted).
- ❌ TypeScript's types give a little to desire.

## [`lavacord`](https://www.npmjs.com/package/lavacord)

- 💚 Explicit connection.
- 💚 Handles player's memory for you.
- 💚 Easy to handle events.
- 🤍 TypeScript's types are decent, but outdated.
- ❌ Support is essentially non-existent.
- ❌ Audio playback stops when Skyra restarts.
- ❌ Holds state of stats, which we do not read.
- ❌ No easy support with Redis libraries.

For `@skyra/lavalink`, we'll most likely fork `lavalink` and `lavaqueue` into either one or two packages, the latter being opt-in. The code is very robust, so it's a better start. Once we're familiarized with the internals, we would rewrite some parts (use `discord-api-types` for Discord's types, make connection explicit, try to find out what causes the crash loop...).

Whatever the outcome is, this will make #1284 a lot easier to resolve.

### PR DESCRIPTION
**Pull request by kyranet**
Fixes #1336.
Fixes #1284.

### Commands

- [x] add
- [x] exportqueue
- [x] importqueue
- [x] join
- [x] leave
- [x] pause
- [x] play
- [x] playing
- [x] promote
- [x] queue-clear
- [x] queue
- [x] remove
- [x] repeat
- [x] resume
- [x] seek
- [x] shuffle
- [x] skip
- [x] volume

### Misc

- [x] LavaLink Events
- [x] Audio Events
- ~~[ ] Dashboard~~ - N.A. -> will be done along with the rest of the dashboard rewrite

### CODE REVIEW

File: src/commands/Music/queue.ts [hunk 1]
```diff
@@ -91,24 +96,53 @@ export default class extends MusicCommand {
...
+		const map = new Map<string, TrackInfo>();
+		for (const entry of decodedTracks) {
+			map.set(entry.track, entry.info);
+		}
+
+		return tracks.map((track) => ({ author: track.author, data: map.get(track.track)! }));
+	}
+}
+
+interface DecodedQueueEntry {
```
Comment (vladfrangu) [inline src/commands/Music/queue.ts:145]: Why is this here?

File: src/events/music/musicQueueSync.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import { Queue } from '@lib/audio';
+import { callOnceAsync } from '@lib/misc';
+import { AudioEvent } from '@lib/structures/AudioEvent';
+import { OutgoingWebsocketAction } from '@lib/websocket/types';
+
+export default class extends AudioEvent {
+	public async run(queue: Queue) {
+		const getTracks = callOnceAsync(() => queue.decodedTracks());
+		for (const subscription of this.getWebSocketListenersFor(queue.guildID)) {
+			subscription.send({ action: OutgoingWebsocketAction.MusicSync, data: { tracks: await getTracks() } });
```
Comment (vladfrangu) [inline src/events/music/musicQueueSync.ts]: Why.. aren't you awaiting this on the `const getTracks = await`? It seems..redundant?

File: src/events/music/musicSongPlay.ts [hunk 3]
```diff
@@ -1,38 +1,17 @@
...
-					action: OutgoingWebsocketAction.MusicSongPlay,
-					data: {
-						song: song.toJSON(),
-						queue: manager.queue.map((s) => s.toJSON())
-					}
-				});
-			}
+		const getTracks = callOnceAsync(() => queue.decodedTracks());
+		for (const subscription of this.getWebSocketListenersFor(queue.guildID)) {
+			subscription.send({ action: OutgoingWebsocketAction.MusicSongPlay, data: { status, tracks: await getTracks() } });
```
Comment (vladfrangu) [inline src/events/music/musicSongPlay.ts]: Ditto

File: src/extendables/VoiceChannelExtendables.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+export default class extends Extendable {
+	public constructor(store: ExtendableStore, file: string[], directory: string) {
+		super(store, file, directory, { appliesTo: [VoiceChannel] });
+	}
+
+	// @ts-expect-error 2784 https://github.com/microsoft/TypeScript/issues/36883
+	public get listeners(this: VoiceChannel): string[] {
+		const members: string[] = [];
+		for (const [id, member] of this.members.entries()) {
+			if (member.user.bot || member.voice.deaf) continue;
```
Comment (vladfrangu) [inline src/extendables/VoiceChannelExtendables.ts:13]: Does deaf account for serverDeafen?

File: src/lib/websocket/WebsocketUser.ts [hunk 5]
```diff
@@ -89,29 +89,32 @@ export default class DashboardWebsocketUser {
...
 
-	private handleSubscriptionUpdate(message: IncomingWebsocketMessage) {
+	private async handleSubscriptionUpdate(message: IncomingWebsocketMessage) {
 		if (!message.data.subscription_name || !message.data.subscription_action) return;
 
 		switch (message.data.subscription_action) {
 			case SubscriptionAction.Subscribe:
-				this.handleSubscribeMessage(message);
-				break;
+				return this.handleSubscribeMessage(message);
```
Comment (vladfrangu) [inline src/lib/websocket/WebsocketUser.ts:97]: Why was this changed into a return but the second case wasn't?

File: src/events/music/musicSongSeekUpdate.ts [hunk 6]
```diff
@@ -1,19 +1,13 @@
...
-			floatPromise(this, channel.sendLocale(LanguageKeys.Commands.Music.SeekSuccess, [{ time: position }]));
-		}
-
-		for (const subscription of manager.websocketUserIterator()) {
-			subscription.send({ action: OutgoingWebsocketAction.MusicSongSeekUpdate, data: { position } });
+export default class extends AudioEvent {
+	public async run(queue: Queue) {
+		const getStatus = callOnceAsync(() => queue.nowPlaying());
+		for (const subscription of this.getWebSocketListenersFor(queue.guildID)) {
+			subscription.send({ action: OutgoingWebsocketAction.MusicSongSeekUpdate, data: { status: await getStatus() } });
```
Comment (vladfrangu) [inline src/events/music/musicSongSeekUpdate.ts]: Ditto..

File: src/events/music/musicSongReplay.ts [hunk 7]
```diff
@@ -1,12 +1,13 @@
...
 
-export default class extends Event {
-	public run(manager: MusicHandler, song: Song) {
-		for (const subscription of manager.websocketUserIterator()) {
-			subscription.send({ action: OutgoingWebsocketAction.MusicSongReplay, data: { song: song.toJSON() } });
+export default class extends AudioEvent {
+	public async run(queue: Queue, status: NP) {
+		const getTracks = callOnceAsync(() => queue.decodedTracks());
+		for (const subscription of this.getWebSocketListenersFor(queue.guildID)) {
+			subscription.send({ action: OutgoingWebsocketAction.MusicSongReplay, data: { status, tracks: await getTracks() } });
```
Comment (vladfrangu) [inline src/events/music/musicSongReplay.ts]: And again

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: src/extendables/VoiceChannelExtendables.ts (hunk 4 above)
Comment (kyranet) [reply src/extendables/VoiceChannelExtendables.ts:13]: Yes https://github.com/discordjs/discord.js/blob/master/src/structures/VoiceState.js#L102-L109.

File: src/commands/Music/queue.ts (hunk 1 above)
Comment (kyranet) [reply src/commands/Music/queue.ts:145]: What do you mean? It's specific to the `queue` command, which is why it's not `export`ed.

**Review by vladfrangu — DISMISSED** [review]

**Review by favna — APPROVED** [review]
